### ISSUE TITLE
stderr is not written to a file when running experiments

### ISSUE THREAD

**Issue by speckdavid** [prost-planner/prost#82]
We should write stderr to a file. Previously we wrote stderr of the rddl server and the prost planner to separate files. However, I think it is best to write both into one file (run.err), as lab is checking this file.

The driver.log file is also not written. We should at least create a dummy file so that the lab doesn't complain.

**Comment by speckdavid** [comment]
There are std.err and std.log files in the testbed folder. These should be rewritten to slurm.err and slurm.log.

**Comment by thomaskeller79** [comment]
In experiment.py, there are files for run.log, run.err, server.log, server.err and driver.log and they are used in the template that starts a process. Is it possible that the files in the testbed folder are files created by the experiment.py script itself?

**Comment by speckdavid** [comment]
The files in the testbed folder correspond to Slurm output, such as aborted jobs due to time/storage limits. However, there are also some messages that refer to segfaults in runs. In standard lab running FD experiments these outputs are written to slurm.out and slum.err. I think we should do something similar.

I have also checked the run.err files, and some are not empty. My current hypothesis is that in the cases where runs are not completed and run.err is empty, we had some major errors (e.g. segfaults errors) which are written to "slurm.err". I will verify this by adding a dummy error message to the prost planner. 

I also noticed that currently the parser_in files are written to the testbed folder, which are not deleted in cases where the run is not completed. We should fix this as well. The best option is probably to start the runs inside the particular run folder so that these files are in the right place.

**Comment by speckdavid** [comment]
The output of prost is correctly written to the file "run.err". However, certain error messages are written to the error file of slurm, which was located in the testbed folder. I also realized that the experiment.py srcipt is only executable from the testbed folder itself and all "parser_in*" files are written to the testbed folder. I have fixed the problems as follows.

- slurm.out and slurm.err are now written to the experiment folder.
- Within the experiment.py and job file, absolute paths are now used by using `os.path.dirname(os.path.abspath(__file__))`.
- Each run is now started from the specific run_dir so that all files are located in the appropriate folder.
- Creating symlinks to the parser to allow the parser to run within a specific run_dir.

This should fix a few problems until we experiment with the Lab framework (#83)

**Comment by thomaskeller79** [comment]
With Python 3.7.3, I get the following error message when I try to execute experiment.py:

```
File "./experiment.py", line 80, in <module>
name = 'prost_' + revision
TypeError: can only concatenate str (not "bytes") to str
```

Is this a Python 3 specific error (which, if I am not mistaken, we should look at in a separate issue) or something that should be fixed here?

**Comment by speckdavid** [comment]
I just had a look at it. Only two minor changes, where necessary, to make Python 3 possible.

**Comment by thomaskeller79** [comment]
Then let's get this out of the way here. Can you make the changes?

**Comment by speckdavid** [comment]
Sure! I pushed it. Does it work now?

**Comment by thomaskeller79** [comment]
The experiment started successfully. I'm waiting for the experiment to finish to see what changed.

**Comment by speckdavid** [comment]
Is there a code formatting standard for Python code that we use?

**Comment by geisserf** [comment]
I don't think we have any official guidelines here. I would stick to similar guidelines that we have for the c++ part (e.g. 80 chars line width), as most of the ones that transfer to python are also PEP8 guidelines. Personally, I use [autopep8](https://pypi.org/project/autopep8/) and have a vim keybind to automatically format the code according to pep8 guidelines.

**Comment by abcorrea** [comment]
We had a similar problem during the IPC. The issue at the time was that we were setting a Slurm memory limit and not a ```ulimit```. If the Slurm limit is reached, then Slurm kills the process and it also produces its output error. Ideally, we should have a soft ulimit (to allow graceful exit of processes) and a Slurm limit above that. But I think there is no ```ulimit``` setting right now in the experiment. 

To see an example, check ```lab```'s  ```lab/data/slurm-job-header.template``` and the definition of ```soft_memory_limit``` in ```lab/environments.py```

If we simply want to output these messages to a same file in the run directory (instead of ```testbed```) we simply need to add

```
### Redirect stdout and stderr.
#SBATCH --output=/path/to/run/slurm.err
#SBATCH --error=/path/to/run/slurm.out
```
to our template.

**Comment by thomaskeller79** [comment]
@abcorrea are you saying that adding these lines:

# Set a soft memory limit to guard against the cgroup mechanism failing (see SlurmEnvironment docs).
ulimit -Sv %(soft_memory_limit)d

and setting the soft_memory_limit to 98% of the slurm memory limit will resolve the issue?

**Comment by abcorrea** [comment]
I think this might solve it. Although I think that 2% is not too much, but it seems to work in Fast Downward.

**Comment by thomaskeller79** [comment]
I wanted to hijack this issue because I thought it was blocking further changes to experiment.py. However, I realized that most of the changes of @speckdavid were already pushed to the repo with the commit we tried to take back (but apparently didn't succeed). I therefore merged the last two lines of code (that were only relevant for python3 support) and decided to start a new issue for the remaining problems with the experiment.py script.
